_posts/2020-06-10-connect-kafka-spring-kafka.md

```diff
@@ -0,0 +1,575 @@
+      }
+    });
+    return kafkaTemplate;
+  }
+}
+```
+
+We can configure KafkaTemplate with a `ProducerListener` (`onSuccess` or `onError`) by using `kafkaTemplate.setProducerListener(producerListener)` in `KafkaProducerConfig` instead of waiting for the `Future` as shown in the above example.
+
+### Using `RoutingKafkaTemplate`
+
+We can use `RoutingKafkaTemplate` when we have multiple producers with different configuration and we want to select producer at runtime based on the topic name.
+
+```java
+@Configuration
+class KafkaProducerConfig {
+  ...
+
+  @Bean
+  public RoutingKafkaTemplate routingTemplate(GenericApplicationContext context) {
+    // ProducerFactory with Bytes serializer
+    Map<String, Object> props = new HashMap<>();
+    props.put(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, bootstrapServers);
+    props.put(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG, StringSerializer.class);
+    props.put(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, ByteArraySerializer.class);
+    DefaultKafkaProducerFactory<Object, Object> bytesPF = new DefaultKafkaProducerFactory<>(props);
+    context.registerBean(DefaultKafkaProducerFactory.class, "bytesPF", bytesPF);
+
+    // ProducerFactory with String serializer
+    props.put(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, StringSerializer.class);
+    DefaultKafkaProducerFactory<Object, Object> stringPF = new DefaultKafkaProducerFactory<>(props);
+
+
+    Map<Pattern, ProducerFactory<Object, Object>> map = new LinkedHashMap<>();
+    map.put(Pattern.compile(".*-bytes"), bytesPF);
+    map.put(Pattern.compile("reflectoring-.*"), stringPF);
+    return new RoutingKafkaTemplate(map);
+  }
+  ...
+}
+```
+
+As shown in the above example, `RoutingKafkaTemplate` takes a map of `java.util.regex.Pattern` and `ProducerFactory<Object, Object>` instances and it is traversed in order, so it is advised to add specific patterns at the beginning (For example: Let's consider we have two patterns `ref.*` and `reflectoring-.*`. In this case `reflectoring-.*` should be at the beginning because `ref.*` pattern also matches the character in reflectoring).
+
+In the above example, we have created two patterns `.*-bytes` and `reflectoring-.*`. The topic names ending with '`-bytes`' and `reflectoring-.*` will use `ByteArraySerializer` and `StringSerializer` respectively when we use `RoutingKafkaTemplate` instance.
+
+## Consuming Records
+
+### Message Listener
+
+As said earlier, `KafkaMessageListenerContainer` receives all records from all topics on a single thread. The `ConcurrentMessageListenerContainer` assigns these records to multiple `KafkaMessageListenerContainer` instances to provide the multi-threaded capability. It also has a concurrency property (`setConcurrency`) which helps to instantiate multiple `KafkaMessageListenerContainer` instances. Let's now look at how to create listeners:
```

Comment:
I removed the part about `setConcurrency` because it's confusing to mention it here and then not explain it in code.